sys-utils/nsenter.c

```diff
@@ -624,32 +728,20 @@ int main(int argc, char *argv[])
 	}
 
 	/*
-	 * Now that we know which namespaces we want to enter, enter
-	 * them.  Do this in two passes, not entering the user
-	 * namespace on the first pass.  So if we're deprivileging the
-	 * container we'll enter the user namespace last and if we're
-	 * privileging it then we enter the user namespace first
-	 * (because the initial setns will fail).
+	 * Now that we know which namespaces we want to enter, enter them.  Do
+	 * this in two passes, not entering the user namespace on the first
+	 * pass.  So if we're deprivileging the container we'll enter the user
+	 * namespace last and if we're privileging it then we enter the user
+	 * namespace first (because the initial setns will fail).
 	 */
-	for (pass = 0; pass < 2; pass ++) {
-		for (nsfile = namespace_files + 1 - pass; nsfile->nstype; nsfile++) {
-			if (nsfile->fd < 0)
-				continue;
-			if (nsfile->nstype == CLONE_NEWPID && do_fork == -1)
-				do_fork = 1;
-			if (setns(nsfile->fd, nsfile->nstype)) {
-				if (pass != 0)
-					err(EXIT_FAILURE,
-					    _("reassociate to namespace '%s' failed"),
-					    nsfile->name);
-				else
-					continue;
-			}
-
-			close(nsfile->fd);
-			nsfile->fd = -1;
-		}
-	}
+	enter_namespaces(pid_fd, namespaces & ~CLONE_NEWUSER, 1);	/* ignore errors */
```

Comment:
I don't think it's necessary to remove CLONE_NEWUSER when using pidfd. It shouldn't cause permission issues when it would work otherwise, but it's a bit wasteful.

Reply:
Kernel evaluates the user namespace as the first one if used with pidfd, so if we want to keep it in the same way as now, it's better to make the first attempt without the user namespace.